### ISSUE TITLE
Create static ToC for Html output

### ISSUE THREAD

**Issue by opoudjis** [metanorma/isodoc#79]

**Comment by opoudjis** [comment]
A static ToC means we give up on the smoothScrooling functionality, of synchronising the scrolling of the ToC and the content window. If you can run the Javascript to do smoothScrooling, you can also run the Javascript the generate the table of contents dynamically. 

Because I will not code Javascript, if that's a problem, I'm not going to be the one to solve it.

**Comment by ronaldtse** [comment]
I think it's fine to first create the static ToC , and have someone else code up the smooth scrolling. @strogonoff could you help here?

**Comment by opoudjis** [comment]
That's not the only Javascript you may need to deal with, if you want to use crippled obsolete junk versions of browsers like Internet Explorer 11. There's also the scroll to the top HTML button, and the rendering of footnotes inline. See https://github.com/riboseinc/metanorma-iso/blob/master/lib/isodoc/iso/html/scripts.html

**Comment by ronaldtse** [comment]
Point remains that our HTML output needs to degrade gracefully. I believe @strogonoff could handle this.

**Comment by opoudjis** [comment]
I'm asking that any Javascript be included in the HTML file; I will add it back into the gem when done.

**Comment by strogonoff** [comment]
@opoudjis are you asking me to send you patched HTML, and you’ll diff & commit the changes later?

**Comment by opoudjis** [comment]
Yes.

**Comment by opoudjis** [comment]
@strogonoff Any progress on this? I am way overcommitted on other projects, but this does still need to be done.

**Comment by opoudjis** [comment]
@strogonoff Re-ping.

**Comment by opoudjis** [comment]
@strogonoff Please action. The task is to extract from the existing JS we use a smaller subset, which will just do mouseover highlighting and smooth scrolling; the remaining ToC functionality will be moving into statically generated code.

**Comment by strogonoff** [comment]
> @strogonoff Please action. The task is to extract from the existing JS we use a smaller subset, which will just do mouseover highlighting and smooth scrolling; the remaining ToC functionality will be moving into statically generated code.

Sure, this slipped through the cracks, let me put some resources on it

**Comment by strogonoff** [comment]
As I’m looking into it, I don’t believe it’s worth mucking around with the jQuery plugin, since generating ToC from headers is its core feature. May be more straightforward to just generate ToC HTML and have a simple script to handle some UX niceties.

@opoudjis, you wrote

> I've generated the static ToC

Is this in a branch? How can I generate the sample document with the static ToC? So that I can play around with the JS side of things on it

**Comment by opoudjis** [comment]
>@opoudjis, you wrote

>>I've generated the static ToC
>Is this in a branch? How can I generate the sample document with the static ToC? So that I can play around with the JS side of things on it

It's commented out code in isodoc:

isodoc/lib/isodoc/html_function/html.rb

````
    def html_toc(docxml)
=begin
      idx = docxml.at("//div[@id = 'toc']") or return docxml
      toc = "<ul>"
      tocidx = 0
      docxml.xpath("//main//h1 | //main//h2[not(@class = 'TermNum')]").each do |h|
        h["id"] ||= "toc#{tocidx}"
        toc += html_toc_entry(h.name == "h1" ? 1 : 2, h)
        tocidx += 1
      end
      idx.children = "#{toc}</ul>"
=end
      docxml
    end
````

Note that the XPath selector which headings to include in the ToC is now being handled by this supplied piece of Javascript, generated dynamically by Ruby:

````
    def toclevel
      <<~HEAD.freeze
    function toclevel() { var i; var text = "";
      for(i = 1; i <= #{@htmlToClevels}; i++) {
        if (i > 1) { text += ","; } text += "h" + i + ":not(:empty):not(.TermNum)"; } 
      return text;}
      HEAD
    end
````

That is in effect a dynamic version of `"//main//h1 | //main//h2[not(@class = 'TermNum')]"` above, which now excludes empty bodies as well as TermNum.

I think you and Ronald need to discuss the point in generating Static ToC for HTML at all, given that Ronald now is talking about bundling jQuery with the HTML code in https://github.com/metanorma/metanorma/issues/59

This is a ticket that has not been actioned in a year, but it is pointless to action it if the distribution strategy for HTML output is going to change so radically.

Which I myself think is insane, but that's now a discussion between @ronaldtse and you, @strogonoff. @ronaldtse, be clear on what you are trying to achieve here, because I think  https://github.com/metanorma/metanorma/issues/59 (if you're going to implement this) renders this ticket obsolete.

**Comment by strogonoff** [comment]
After chat with Ronald, I’ll get static ToC done. jQuery may still need to be bundled, even if new ToC doesn’t use it other parts would.

**Comment by strogonoff** [comment]
@opoudjis I may be missing something, but the document works with static ToC as is.

Meaning:

— With code commented and JS disabled, ToC doesn’t work (as expected)
— With code uncommented and JS disabled, ToC works (except for highlighting current item, also as expected)
— With code uncommented and JS enabled, ToC behavior is identical

Could you clarify why the script may need to be modified?

**Comment by strogonoff** [comment]
Tested on `isodoc-rice` & Safari. Is there a browser-specific issue where the static ToC breaks because of that jQuery plugin?

**Comment by strogonoff** [comment]
@opoudjis To be more precise:

> mouseover does highlight items, but it is not sticky: on click, the class "toc-active" ("activeClass" within the script) needs to be added to the clicked list item.

This works for me the same with or without static ToC.

> the functionality of smoothScrolling, which automatically scrolls the sidebar when the main window is scrolled, needs to be replicated.

This does _not_ work for me, again with or without static ToC.

**Comment by opoudjis** [comment]
Huh. These are indeed broken. Ok, I'll regenerate, hold on.

**Comment by opoudjis** [comment]
OK, sorry. This is what's meant to happen:

* The version of the HTML I sent you was the one without highlighting of the current ToC item. Use the attached instead.
* The expected smoothscrolling behaviour in the given script is an animation of scrolling, not synchronised scrolling, so you can ignore it. I am convinced that I have seen synchronised scrolling in the past, but I can't replicate it now.

The missing activeClass functionality is:

````
j = function(b, c) {
                /* ... */
                a("li", d).removeClass(i), a(b.target).parent().addClass(i)
            },
````

so basically, function("onclick", e.addClass("toc-active").

Rather than play middleman about dodgy functionality in a 10 year old ToC script, I invite @ronaldtse to communicate to you direct. You now have a dynamic and a static version of the ToC page attached here. Ronald, you tell Anton what you see missing (apart from the on-click class functionality.)

[rice_alt.html.dynamic.zip](https://github.com/metanorma/isodoc/files/3444634/rice_alt.html.dynamic.zip)
[rice_alt.html.static.zip](https://github.com/metanorma/isodoc/files/3444635/rice_alt.html.static.zip)

**Comment by opoudjis** [comment]
So the original scope of the ticket is now minimal: it's merely giving me a one line javascript command to change css class on click.

But the potential scope of this ticket is a lot bigger: implement whatever @ronaldtse would like to make the HTML ToC pretty, including simultaneous scrolling.

When you've worked out what the JS should be, let me know, and I will add it to the isodoc template.

**Comment by strogonoff** [comment]
Nick, the highlight of the active item in sidebar works for me with the uncommented version & ToC fully static, with existing JS. I’ll attach a screencast.

**Comment by strogonoff** [comment]
@opoudjis Here’s what I see. First on the left is the original (without static ToC), then on the right the version with uncommented code you pointed to (with static ToC). No JS modifications made so far, only uncommented your code.

![Jul-30-2019 15-06-24](https://user-images.githubusercontent.com/292959/62104870-d0ab9b00-b2db-11e9-89b2-056dcc39ee98.gif)

 Active item highlight works on both. Scrolling sync does not work on either.

**Comment by opoudjis** [comment]
Scrolling sync indeed does not exist anywhere, it was a feature request by Ronald. 

You are looking at the .html file in the screenshot. Look at the .alt.html file,  which highlights the clicked on TOC item by adding a CSS class to it. The static version can't be doing it, because I've commented out the TOC script that does that.

**Comment by strogonoff** [comment]
@opoudjis I am not looking at your attachments, I am looking at the document I built myself as described [here](https://www.metanorma.com/author/iso/sample/). I built it two times, one with uncommented code in the gem. I don’t see any issues.

**Comment by opoudjis** [comment]
Uncommenting the isodoc code is not sufficient. The point of uncommenting the isodoc code is that we should no longer need to call the ToC script, as is done e.g. in the first script in, metanorma-iso/lib/isodoc/iso/html/scripts.html, `$('#toc').toc({.....})`

If you do not remove the ToC script call from the HTML file, then the ToC script still ends up being run, and building the ToC dynamically. Which means there is no point in generating the ToC statically. That script also highlights clicked ToC entries, by assigning a CSS class to it. 

If you do remove the ToC script call from the HTML file, then there is nothing to highlight the clicked on ToC entry. I don't need help doing that, that is a one-line JS function.

However, your task is to liaise with Ronald, work out if *anything else* needs to accompany the ToC as a javascript script (especially his request for synchronised scrolling), and give me whatever is needed as a new javascript script, which does not construct the ToC dynamically, but which does do whatever other rendering control is needed.

**Comment by strogonoff** [comment]
> If you do not remove the ToC script call from the HTML file, then the ToC script still ends up being run, and building the ToC dynamically. Which means there is no point in generating the ToC statically. That script also highlights clicked ToC entries, by assigning a CSS class to it.

I disagree:

1. The point of having static ToC is to make the document degrade gracefully.
2. The point of having dynamic ToC is to retain the dynamic ToC functionality.
3. Leaving both on is the quickest way to resolve the actual problem the customer is facing, causing no difference in behavior perceptible by the customer.

A proper longer-term solution can follow this. However, I believe we should address the problem ASAP and then refactor if we want to.

**Comment by strogonoff** [comment]
I think we agree on principle, I just suggest that we uncomment the static ToC as soon as customer confirms it solves their problem (dynamic ToC doesn’t break static ToC from my tests), and later refactor & improve on that (filed https://github.com/metanorma/isodoc/issues/113 for that)

**Comment by opoudjis** [comment]
Agree with you, I've discussed this independently with Ronald. So, I am now going to uncomment the isodoc code, and make sure it responds to gem-specific settings of permitted ToC styles.

**Comment by strogonoff** [comment]
It looks like html_toc function in this `isodoc` repo doesn’t affect `metanorma-ogc` deliverables. I believe the primary complaint was from the OGC guys. I’m not sure whether we should reopen this or create another one in OGC

**Comment by opoudjis** [comment]
No, under this ticket, I need to remove all overriding html_toc function from gems, I hadn't realised they had been there (effectively disabling static HTML TOC).
